File: app/src/main/java/com/polok/eubmanagement/presentation/assignment/assignmentadd/AssignmentAddViewModel.java
package com.polok.eubmanagement.presentation.assignment.assignmentadd;

import static com.polok.eubmanagement.util.Extension.showErrorOnUi;
import android.widget.EditText;
import androidx.annotation.NonNull;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.database.DatabaseReference;
import com.polok.eubmanagement.base.BaseViewModel;
import com.polok.eubmanagement.firebase.FirebaseDataRef;
import com.polok.eubmanagement.presentation.assignment.modal.AssignmentData;
import com.polok.eubmanagement.util.Extension;

public class AssignmentAddViewModel extends BaseViewModel {

    public void validateAssignmentInputAndUploadToFirebase(
            EditText assignmentTitleEt, EditText assignmentDetailsEt
    ) {
        if (assignmentTitleEt.getText().toString().isEmpty()) {
            showErrorOnUi(assignmentTitleEt, "Enter Assignment Title");
            return;
        }
        if (assignmentDetailsEt.getText().toString().isEmpty()) {
            showErrorOnUi(assignmentDetailsEt, "Enter Assignment Details");
            return;
        }
        uploadAssignmentTOFirebase(
                assignmentTitleEt.getText().toString(),
                assignmentDetailsEt.getText().toString(),
                Extension.getCurrentDate()
        );
    }

    private void uploadAssignmentTOFirebase(String title, String details, String date) {
        fireLoadingEvent(true);
        DatabaseReference pushAssignmentRef = FirebaseDataRef.provideAssignmentRef().push();
        pushAssignmentRef.setValue(
                new AssignmentData(title,details,date,pushAssignmentRef.getKey().toString())
        ).addOnCompleteListener(new OnCompleteListener<Void>() {
            @Override
            public void onComplete(@NonNull Task<Void> task) {
                if (task.isComplete()) {
                    fireMessageEvent("Assignment Added Successfully");
                    fireNavigateEvent(0,null);
                } else fireMessageEvent(task.getException().getLocalizedMessage());
                fireLoadingEvent(false);
            }
        });
    }
}

File: app/src/main/java/com/polok/eubmanagement/presentation/assignment/assignmentadd/AssignmentAddViewModel.kt
package com.polok.eubmanagement.presentation.assignment.assignmentadd

import android.widget.EditText
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.polok.eubmanagement.base.BaseViewModel
import com.polok.eubmanagement.firebase.FirebaseDataRef.provideAssignmentRef
import com.polok.eubmanagement.model.AssignmentData
import com.polok.eubmanagement.util.getCurrentDate
import com.polok.eubmanagement.util.showErrorOnUi

class AssignmentAddViewModel : BaseViewModel() {

    fun validateAssignmentInputAndUploadToFirebase(
        assignmentTitleEt: EditText, assignmentDetailsEt: EditText
    ) {
        if (assignmentTitleEt.getText().toString().isEmpty()) {
            assignmentTitleEt.showErrorOnUi("Enter Assignment Title")
            return
        }
        if (assignmentDetailsEt.getText().toString().isEmpty()) {
            assignmentDetailsEt.showErrorOnUi("Enter Assignment Details")
            return
        }
        uploadAssignmentTOFirebase(
            assignmentTitleEt.getText().toString(),
            assignmentDetailsEt.getText().toString(),
            getCurrentDate()
        )
    }

    private fun uploadAssignmentTOFirebase(title: String, details: String, date: String) {
        fireLoadingEvent(true)
        val dbPushRes = provideAssignmentRef()?.push()
        dbPushRes?.setValue(
            AssignmentData(
                title = title,
                details = details,
                createdAt = date,
                key = dbPushRes.key
            )
        )?.addOnCompleteListener { task ->
            if (task.isComplete) {
                fireMessageEvent("Assignment Saved Successfully")
                fireNavigateEvent(0, null)
            } else {
                fireMessageEvent(task.exception?.localizedMessage)
            }
            fireLoadingEvent(false)
        }
    }

    class Factory : ViewModelProvider.Factory {
        @Suppress("UNCHECKED_CAST")
        override fun <T : ViewModel> create(modelClass: Class<T>): T {
            return AssignmentAddViewModel() as T
        }
    }
}